File: client-sdk/typescript/src/identity/signer.ts
// Signing capability for the sender-identity extension.
//
// Both SDKs take a pre-opened connection and never see its credentials, so
// the seed is handed in explicitly through a `SenderSigner`. The built-in
// implementation wraps an `@nats-io/nkeys` user key pair; users with an
// HSM / KMS implement the interface themselves (`sign` may be async).
//
// Key hygiene: the key pair lives in a private field, is redacted from
// `JSON.stringify` and `util.inspect`, the seed text is not retained, and
// `wipe()` clears the pair. Error messages never include input text —
// public keys and line numbers only.
//
// No `@nats-io/jwt`: creds files and JWT payloads are parsed by hand (the
// `-----BEGIN NATS USER JWT-----` / `-----BEGIN USER NKEY SEED-----` blocks,
// and a base64url JSON payload decode — no signature check, the server
// already authenticated the JWT).

import { readFile } from "node:fs/promises";
import { fromSeed, type KeyPair } from "@nats-io/nkeys";
import { expandHome, readAuthFile, readContextFile } from "../context.js";
import { IdentityError, IdentityMismatchError, IdentityUnavailableError } from "../errors.js";
import { agentIdUser, newAgentId, type AgentId } from "./agent-id.js";
import { base64UrlDecode, utf8 } from "./crypto.js";

/** Something that can sign with the connection's user NKEY seed. */
export interface SenderSigner {
  /** The user public NKEY (`U…`) this signer signs for. */
  readonly publicKey: string;
  /**
   * The user JWT, when the signer came from a credentials file. Live
   * connection binding compares its user and account with `$SYS.REQ.USER.INFO`.
   */
  readonly jwt?: string;
  /** ed25519 signature over `data`. May be async (HSM / KMS signers). */
  sign(data: Uint8Array): Uint8Array | Promise<Uint8Array>;
  /** Clear the key material; later `sign()` calls fail. Optional for custom signers. */
  wipe?(): void;
}

const USER_SEED_REGEX = /^SU[A-Z2-7]{56}$/;
const JWT_REGEX = /^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]*$/;
const BEGIN_JWT_REGEX = /^-+BEGIN NATS USER JWT-+$/;
const BEGIN_SEED_REGEX = /^-+BEGIN USER NKEY SEED-+$/;

class NkeySigner implements SenderSigner {
  #kp: KeyPair | null;
  readonly publicKey: string;
  readonly jwt?: string;

  constructor(kp: KeyPair, jwt: string | undefined) {
    this.#kp = kp;
    this.publicKey = kp.getPublicKey();
    if (jwt !== undefined) this.jwt = jwt;
  }

  sign(data: Uint8Array): Uint8Array {
    if (this.#kp === null) throw new IdentityError("signer has been wiped");
    return this.#kp.sign(data);
  }

  wipe(): void {
    this.#kp?.clear();
    this.#kp = null;
  }

  toJSON(): { publicKey: string } {
    return { publicKey: this.publicKey };
  }

  [Symbol.for("nodejs.util.inspect.custom")](): string {
    return `SenderSigner(${this.publicKey})`;
  }

  toString(): string {
    return `SenderSigner(${this.publicKey})`;
  }
}

/**
 * Build a signer from a user seed (`SU…`), given as text or bytes.
 * Surrounding whitespace (a trailing newline from a seed file) is ignored,
 * and a `-----BEGIN USER NKEY SEED-----` block is accepted too. The seed
 * is not retained; the message of a rejection never includes it.
 */
export function signerFromSeed(seed: string | Uint8Array, jwt?: string): SenderSigner {
  const canonicalSeed = normalizeUserSeed(seed);
  return signerFromCanonicalSeed(canonicalSeed, jwt);
}

/** Use an already-normalized user-seed snapshot without copying it. @internal */
export function signerFromCanonicalSeed(canonicalSeed: Uint8Array, jwt?: string): SenderSigner {
  // `fromSeed` keeps this buffer by reference and decodes it lazily on
  // every `sign()` / `getPublicKey()`. Ownership transfers to the key pair;
  // `wipe()` -> `kp.clear()` is what zeroes it.
  try {
    return new NkeySigner(fromSeed(canonicalSeed), jwt);
  } catch (err) {
    canonicalSeed.fill(0);
    throw new IdentityError(
      `invalid nkey seed (${err instanceof Error ? err.message : String(err)})`,
    );
  }
}

/**
 * Normalize a raw seed file / seed block to canonical `SU...` bytes, and
 * validate both its checksum and user-key prefix. Internal connection-bundle
 * seam so the NATS authenticator and sender signer can share one exact byte
 * snapshot.
 *
 * @internal
 */
export function normalizeUserSeed(seed: string | Uint8Array): Uint8Array {
  const text = typeof seed === "string" ? seed : utf8.decode(seed);
  const line = text.includes("BEGIN")
    ? extractBlock(text, BEGIN_SEED_REGEX, "USER NKEY SEED")
    : text.trim();
  if (!USER_SEED_REGEX.test(line)) {
    throw new IdentityError("invalid nkey seed: expected a user seed (SU + 56 base32 characters)");
  }
  const canonicalSeed = utf8.encode(line);
  // `KeyPair.clear()` zeroes the input buffer retained by `fromSeed`, so use
  // a throwaway copy for validation and return a separate canonical snapshot.
  const validationSeed = canonicalSeed.slice();
  let kp: KeyPair | undefined;
  let valid = false;
  try {
    kp = fromSeed(validationSeed);
    // nkeys decodes lazily, so validation includes getPublicKey(), not only
    // fromSeed(). This is where a syntactically valid seed with a bad CRC
    // fails.
    const publicKey = kp.getPublicKey();
    if (!publicKey.startsWith("U")) {
      throw new IdentityError(
        `seed is not a user seed (derives ${publicKey.slice(0, 1)}… public key)`,
      );
    }
    valid = true;
    return canonicalSeed;
  } catch (err) {
    if (err instanceof IdentityError) throw err;
    throw new IdentityError(
      `invalid nkey seed (${err instanceof Error ? err.message : String(err)})`,
    );
  } finally {
    kp?.clear();
    validationSeed.fill(0);
    if (!valid) canonicalSeed.fill(0);
  }
}

/** The two blocks of a credentials file. */
export interface ParsedCreds {
  readonly jwt: string;
  readonly seed: string;
}

/**
 * Hand-rolled `.creds` parser: the line after `-----BEGIN NATS USER JWT-----`
 * is the JWT, the line after `-----BEGIN USER NKEY SEED-----` is the seed.
 * Error messages carry line numbers, never content.
 */
export function parseCreds(text: string): ParsedCreds {
  const jwt = extractBlock(text, BEGIN_JWT_REGEX, "NATS USER JWT");
  if (!JWT_REGEX.test(jwt)) throw new IdentityError("creds: the user JWT block is not a JWT");
  const seed = extractBlock(text, BEGIN_SEED_REGEX, "USER NKEY SEED");
  return { jwt, seed };
}

function extractBlock(text: string, begin: RegExp, label: string): string {
  const lines = text.split(/\r?\n/);
  for (let i = 0; i < lines.length; i++) {
    if (!begin.test(lines[i]!.trim())) continue;
    for (let j = i + 1; j < lines.length; j++) {
      const candidate = lines[j]!.trim();
      if (candidate.length === 0) continue;
      if (candidate.startsWith("-")) {
        throw new IdentityError(`creds: empty ${label} block at line ${i + 1}`);
      }
      return candidate;
    }
    throw new IdentityError(`creds: ${label} block at line ${i + 1} has no content`);
  }
  throw new IdentityError(`creds: no -----BEGIN ${label}----- block found`);
}

/** Decode the payload (second part) of a JWT without verifying it. */
export function decodeJwtPayload(jwt: string): Record<string, unknown> {
  const parts = jwt.split(".");
  if (parts.length !== 3) throw new IdentityError("JWT does not have three parts");
  let payload: unknown;
  try {
    payload = JSON.parse(utf8.decode(base64UrlDecode(parts[1]!.replace(/=+$/, ""))));
  } catch {
    throw new IdentityError("JWT payload is not base64url-encoded JSON");
  }
  if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
    throw new IdentityError("JWT payload is not a JSON object");
  }
  return payload as Record<string, unknown>;
}

/**
 * The agent ID a user JWT carries: `sub` is the user NKEY; the account is
 * `nats.issuer_account` when a signing key issued the JWT, else `iss`.
 */
export function identityFromJwt(jwt: string): AgentId {
  const payload = decodeJwtPayload(jwt);
  const sub = payload["sub"];
  const iss = payload["iss"];
  const nats = payload["nats"];
  const issuerAccount =
    typeof nats === "object" && nats !== null && !Array.isArray(nats)
      ? (nats as Record<string, unknown>)["issuer_account"]
      : undefined;
  const account =
    typeof issuerAccount === "string" && issuerAccount.length > 0 ? issuerAccount : iss;
  if (typeof sub !== "string" || typeof account !== "string") {
    throw new IdentityUnavailableError("credentials JWT lacks a string `sub` or `iss`");
  }
  try {
    return newAgentId(account, sub);
  } catch (err) {
    throw new IdentityUnavailableError(
      `credentials JWT does not carry a usable (account, user) pair: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}

/**
 * Build a signer from credentials-file text. The signer carries the user
 * JWT so live connection binding can compare its user and account with
 * `$SYS.REQ.USER.INFO`. A
 * seed that does not belong to the JWT's `sub` is rejected with
 * `IdentityMismatchError` (public keys only in the message).
 */
export function signerFromCreds(credsText: string): SenderSigner {
  const { jwt, seed } = parseCreds(credsText);
  return signerFromSeedAndJwt(seed, jwt);
}

/**
 * Build a signer from one seed/JWT snapshot and verify that the JWT names
 * the derived user. Internal connection-bundle seam; intentionally not
 * exported from the package root.
 *
 * @internal
 */
export function signerFromSeedAndJwt(seed: string | Uint8Array, jwt: string): SenderSigner {
  return signerFromCanonicalSeedAndJwt(normalizeUserSeed(seed), jwt);
}

/** Verify a JWT against an already-normalized seed snapshot. @internal */
export function signerFromCanonicalSeedAndJwt(
  canonicalSeed: Uint8Array,
  jwt: string,
): SenderSigner {
  const signer = signerFromCanonicalSeed(canonicalSeed, jwt);
  try {
    const jwtUser = agentIdUser(identityFromJwt(jwt));
    if (jwtUser === signer.publicKey) return signer;
    throw new IdentityMismatchError(
      signer.publicKey,
      signer.publicKey,
      undefined,
      undefined,
      jwtUser,
    );
  } catch (error) {
    signer.wipe?.();
    throw error;
  }
}

/** {@link signerFromCreds} over a file path (`~/` expanded). */
export async function signerFromCredsFile(path: string): Promise<SenderSigner> {
  const resolved = expandHome(path);
  let text: string;
  try {
    text = await readFile(resolved, "utf8");
  } catch (error) {
    throw new IdentityError(`failed to read creds file ${resolved}`, { cause: error });
  }
  return signerFromCreds(text);
}

/**
 * Build a signer from a `nats` CLI context (`"current"` resolves the
 * selected one): `creds` → {@link signerFromCredsFile}; else `nkey` (a seed
 * file) → {@link signerFromSeed}; else inline `user_seed` (+ `user_jwt`).
 * Reuses the context *reader*, not the connection-option builder.
 */
export async function signerFromContext(selector: string): Promise<SenderSigner> {
  const { name, fields } = await readContextFile(selector);
  const creds = optionalString(fields["creds"]);
  const nkey = optionalString(fields["nkey"]);
  const userSeed = optionalString(fields["user_seed"]);
  const userJwt = optionalString(fields["user_jwt"]);
  if (creds !== undefined) return signerFromCredsFile(creds);
  if (nkey !== undefined) {
    const bytes = await readAuthFile("nkey", expandHome(nkey));
    return signerFromSeed(new Uint8Array(bytes.buffer, bytes.byteOffset, bytes.byteLength));
  }
  if (userSeed !== undefined) {
    return userJwt === undefined
      ? signerFromSeed(userSeed)
      : signerFromSeedAndJwt(userSeed, userJwt);
  }
  throw new IdentityError(
    `NATS context "${name}" has no creds, nkey, or user_seed — nothing to sign Agent-Sender with`,
  );
}

function optionalString(v: unknown): string | undefined {
  return typeof v === "string" && v.length > 0 ? v : undefined;
}

File: client-sdk/python/src/synadia_ai/agents/identity/signer.py
"""Signing capability for the sender-identity extension.

Both SDKs take a pre-opened connection and never see its credentials, so
the seed is handed in explicitly through a :class:`SenderSigner`. The
built-in implementation wraps an ``nkeys`` user key pair; users with an
HSM / KMS implement the protocol themselves (``sign`` may be async).

Key hygiene: the key pair lives in a private attribute, is redacted from
``repr()`` / ``str()``, the seed buffer is zeroed once the public key is
cached (``nkeys.from_seed`` keeps a reference to the buffer it was given
and re-decodes it only for ``public_key``, which this class caches
first), and :meth:`NkeySigner.wipe` clears the pair. Error messages never
include input text — public keys and line numbers only.

No ``@nats-io/jwt`` equivalent: creds files and JWT payloads are parsed by
hand (the ``-----BEGIN NATS USER JWT-----`` / ``-----BEGIN USER NKEY
SEED-----`` blocks, and a base64url JSON payload decode — no signature
check, the server already authenticated the JWT).
"""

from __future__ import annotations

import base64
import binascii
import inspect
import json
import re
from collections.abc import Awaitable
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Protocol, TypeVar

import nkeys

from ..context import expand_user, read_context_file
from ..errors import IdentityError, IdentityMismatchError, IdentityUnavailableError
from ._nkeys import base64url_decode
from .agent_id import AgentId

T = TypeVar("T")

_USER_SEED_REGEX = re.compile(r"^SU[A-Z2-7]{56}$")
_JWT_REGEX = re.compile(r"^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]*$")
_BEGIN_JWT_REGEX = re.compile(r"^-+BEGIN NATS USER JWT-+$")
_BEGIN_SEED_REGEX = re.compile(r"^-+BEGIN USER NKEY SEED-+$")
_JWT_PARTS = 3
_CRC_BYTES = 2


class SenderSigner(Protocol):
    """Something that can sign with the connection's user NKEY seed.

    ``sign`` may return the signature directly or an awaitable of it
    (HSM / KMS signers). ``jwt`` is the user JWT when the signer came from
    a credentials file — ``self_id()`` reads the agent ID from it (no
    network, not spoofable by peers); ``None`` otherwise.
    """

    @property
    def public_key(self) -> str:
        """The user public NKEY (``U…``) this signer signs for."""
        ...

    @property
    def jwt(self) -> str | None:
        """The user JWT, when the signer came from a credentials file."""
        ...

    def sign(self, data: bytes) -> bytes | Awaitable[bytes]:
        """ed25519 signature over ``data``."""
        ...


async def maybe_await(value: T | Awaitable[T]) -> T:
    """Await ``value`` when it is awaitable (an async ``sign`` / hook), else return it."""
    if inspect.isawaitable(value):
        return await value
    return value


class NkeySigner:
    """The built-in :class:`SenderSigner` over an ``nkeys`` user key pair."""

    __slots__ = ("_jwt", "_kp", "_public_key")

    def __init__(self, kp: nkeys.KeyPair, jwt: str | None) -> None:
        self._kp: nkeys.KeyPair | None = kp
        # Cache before the seed buffer is zeroed: `public_key` re-decodes
        # the seed on first access.
        self._public_key: str = kp.public_key.decode("ascii")
        self._jwt = jwt

    @property
    def public_key(self) -> str:
        return self._public_key

    @property
    def jwt(self) -> str | None:
        return self._jwt

    def sign(self, data: bytes) -> bytes:
        if self._kp is None:
            raise IdentityError("signer has been wiped")
        return bytes(self._kp.sign(data))

    def wipe(self) -> None:
        """Clear the key material; later :meth:`sign` calls fail."""
        if self._kp is not None:
            self._kp.wipe()
            self._kp = None

    def __repr__(self) -> str:
        return f"SenderSigner({self._public_key})"

    __str__ = __repr__


def signer_from_seed(seed: str | bytes, jwt: str | None = None) -> NkeySigner:
    """Build a signer from a user seed (``SU…``), given as text or bytes.

    Surrounding whitespace (a trailing newline from a seed file) is
    ignored, and a ``-----BEGIN USER NKEY SEED-----`` block is accepted
    too. The seed is not retained; the message of a rejection never
    includes it.
    """
    text = seed if isinstance(seed, str) else seed.decode("utf-8", errors="replace")
    line = (
        _extract_block(text, _BEGIN_SEED_REGEX, "USER NKEY SEED")
        if "BEGIN" in text
        else text.strip()
    )
    if _USER_SEED_REGEX.match(line) is None:
        raise IdentityError("invalid nkey seed: expected a user seed (SU + 56 base32 characters)")
    buf = bytearray(line.encode("ascii"))
    try:
        raw = base64.b32decode(bytes(buf) + b"=" * (-len(buf) % 8))
        if nkeys.crc16_checksum(raw[:-_CRC_BYTES]) != raw[-_CRC_BYTES:]:
            raise IdentityError("invalid nkey seed (CRC check failed)")
        kp = nkeys.from_seed(buf)
        signer = NkeySigner(kp, jwt)
    except (nkeys.NkeysError, binascii.Error, ValueError) as exc:
        raise IdentityError(f"invalid nkey seed ({type(exc).__name__})") from exc
    finally:
        # `from_seed` keeps `buf` by reference; the public key is cached
        # above, and signing needs only the derived key — zero the text.
        for i in range(len(buf)):
            buf[i] = 0
    if not signer.public_key.startswith("U"):
        signer.wipe()
        raise IdentityError(
            f"seed is not a user seed (derives {signer.public_key[:1]}… public key)"
        )
    return signer


@dataclass(frozen=True, slots=True)
class ParsedCreds:
    """The two blocks of a credentials file."""

    jwt: str
    seed: str


def parse_creds(text: str) -> ParsedCreds:
    """Hand-rolled ``.creds`` parser.

    The line after ``-----BEGIN NATS USER JWT-----`` is the JWT, the line
    after ``-----BEGIN USER NKEY SEED-----`` is the seed. Error messages
    carry line numbers, never content.
    """
    jwt = _extract_block(text, _BEGIN_JWT_REGEX, "NATS USER JWT")
    if _JWT_REGEX.match(jwt) is None:
        raise IdentityError("creds: the user JWT block is not a JWT")
    seed = _extract_block(text, _BEGIN_SEED_REGEX, "USER NKEY SEED")
    return ParsedCreds(jwt=jwt, seed=seed)


def _extract_block(text: str, begin: re.Pattern[str], label: str) -> str:
    lines = text.splitlines()
    for i, line in enumerate(lines):
        if begin.match(line.strip()) is None:
            continue
        for candidate in (ln.strip() for ln in lines[i + 1 :]):
            if not candidate:
                continue
            if candidate.startswith("-"):
                raise IdentityError(f"creds: empty {label} block at line {i + 1}")
            return candidate
        raise IdentityError(f"creds: {label} block at line {i + 1} has no content")
    raise IdentityError(f"creds: no -----BEGIN {label}----- block found")


def decode_jwt_payload(jwt: str) -> dict[str, Any]:
    """Decode the payload (second part) of a JWT without verifying it."""
    parts = jwt.split(".")
    if len(parts) != _JWT_PARTS:
        raise IdentityError("JWT does not have three parts")
    try:
        payload = json.loads(base64url_decode(parts[1].rstrip("=")).decode("utf-8"))
    except (ValueError, UnicodeDecodeError) as exc:
        raise IdentityError("JWT payload is not base64url-encoded JSON") from exc
    if not isinstance(payload, dict):
        raise IdentityError("JWT payload is not a JSON object")
    return payload


def identity_from_jwt(jwt: str) -> AgentId:
    """The agent ID a user JWT carries.

    ``sub`` is the user NKEY; the account is ``nats.issuer_account`` when
    a signing key issued the JWT, else ``iss``.
    """
    payload = decode_jwt_payload(jwt)
    sub = payload.get("sub")
    iss = payload.get("iss")
    nats_claims = payload.get("nats")
    issuer_account = nats_claims.get("issuer_account") if isinstance(nats_claims, dict) else None
    account = issuer_account if isinstance(issuer_account, str) and issuer_account else iss
    if not isinstance(sub, str) or not isinstance(account, str):
        raise IdentityUnavailableError("credentials JWT lacks a string `sub` or `iss`")
    try:
        return AgentId.new(account, sub)
    except IdentityError as exc:
        raise IdentityUnavailableError(
            f"credentials JWT does not carry a usable (account, user) pair: {exc}"
        ) from exc


def signer_from_creds(creds_text: str) -> NkeySigner:
    """Build a signer from credentials-file text.

    The signer carries the user JWT so ``self_id()`` can read the identity
    without asking the server. A seed that does not belong to the JWT's
    ``sub`` is rejected with :class:`IdentityMismatchError` (public keys
    only in the message).
    """
    parsed = parse_creds(creds_text)
    signer = signer_from_seed(parsed.seed, parsed.jwt)
    jwt_user = identity_from_jwt(parsed.jwt).user
    if jwt_user != signer.public_key:
        signer.wipe()
        raise IdentityMismatchError(signer.public_key, jwt_user)
    return signer


def signer_from_creds_file(path: str | Path) -> NkeySigner:
    """:func:`signer_from_creds` over a file path (``~/`` expanded)."""
    resolved = Path(expand_user(str(path)))
    try:
        text = resolved.read_text(encoding="utf-8")
    except OSError as exc:
        raise IdentityError(f"failed to read creds file {resolved}: {exc.strerror}") from exc
    return signer_from_creds(text)


def signer_from_context(selector: str) -> NkeySigner:
    """Build a signer from a ``nats`` CLI context (``"current"`` resolves the selected one).

    ``creds`` → :func:`signer_from_creds_file`; else ``nkey`` (a seed file)
    → :func:`signer_from_seed`; else inline ``user_seed`` (+ ``user_jwt``).
    Reuses the context *reader*, not the connection-option builder.
    """
    ctx = read_context_file(selector)
    creds = _optional_str(ctx.fields.get("creds"))
    nkey = _optional_str(ctx.fields.get("nkey"))
    user_seed = _optional_str(ctx.fields.get("user_seed"))
    user_jwt = _optional_str(ctx.fields.get("user_jwt"))
    if creds is not None:
        return signer_from_creds_file(creds)
    if nkey is not None:
        seed_path = Path(expand_user(nkey))
        try:
            return signer_from_seed(seed_path.read_bytes())
        except OSError as exc:
            raise IdentityError(
                f"failed to read nkey seed file {seed_path}: {exc.strerror}"
            ) from exc
    if user_seed is not None:
        signer = signer_from_seed(user_seed, user_jwt)
        if user_jwt is not None:
            jwt_user = identity_from_jwt(user_jwt).user
            if jwt_user != signer.public_key:
                signer.wipe()
                raise IdentityMismatchError(signer.public_key, jwt_user)
        return signer
    raise IdentityError(
        f"NATS context {ctx.name!r} has no creds, nkey, or user_seed — nothing to sign "
        "Agent-Sender with"
    )


def _optional_str(value: object) -> str | None:
    return value if isinstance(value, str) and value else None


__all__ = [
    "NkeySigner",
    "ParsedCreds",
    "SenderSigner",
    "decode_jwt_payload",
    "identity_from_jwt",
    "maybe_await",
    "parse_creds",
    "signer_from_context",
    "signer_from_creds",
    "signer_from_creds_file",
    "signer_from_seed",
]
